src/sql/workbench/services/executionPlan/common/executionPlanService.ts

```diff
@@ -27,19 +28,37 @@ export class ExecutionPlanService implements IExecutionPlanService {
 		this._providerRegisterEvent = this._onProviderRegister.event;
 	}
 
-	private async ensureCapabilitiesRegistered(): Promise<void> {
-		let providers = Object.keys(this._capabilitiesService.providers);
-		if (!providers) {
-			await new Promise<void>(resolve => {
-				this._capabilitiesService.onCapabilitiesRegistered(e => {
-					resolve();
-				});
-			});
-		}
+	/**
+	 * This ensures that the capabilities service has registered the providers to handle execution plan requests.
+	 * @param providerId Optional provider id to wait for.
+	 */
+	private async ensureCapabilitiesRegistered(providerId?: string): Promise<void> {
+		let providers: string[] = [];
+		// Wait until the capabilities service has registered some providers.
+		await new Promise<void>(resolve => {
+			let retryCount = 0;
+			const intervalId = setInterval(() => {
```

Comment:
There was a race condition because sometimes the providers were registered before this file was able to resgiter onCapabilitiesRegistered handler. This code fixes that.

Reply:
Then couldn't you just check to see if the provider is registered already and resolve immediately, otherwise set up the promise so it waits for it to resolve?

Reply:
I agree with Charles, this can be simplified. also looks like you are simply returning after retrying for 5 times, are we not supposed to reject the promise with some error message?

Reply:
Those errors are already handled later @alanrenmsft

Reply:
Why are you retrying anyways? We get notified when a provider is registered. It'd be better to just have a single timeout in case it never gets registered.